File: Week_01/G20190343020222/merge-two-sorted-lists.js
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
    if (!l1) {
        return l2;
    }

    if (!l2) {
        return l1;
    }   

    const dummy = new ListNode(-1);
    let cur = dummy;

    while (l1 && l2) {
        if (l1.val <= l2.val) {
            cur.next = l1;
            cur = l1;
            l1 = l1.next
        } else {
            cur.next = l2;
            cur = l2;
            l2 = l2.next;
        }
    }

    if (l1 || l2) {
        cur.next = l1 ? l1 : l2;
    }

    return dummy.next;
};
// @lc code=end

File: Week_01/G20190343020224/MergeTwoSortedLists.java
class MergeTwoSortedLists {

    //暴力
    class Solution {
        public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
            ListNode head = new ListNode(0);
            ListNode cur = head;
            while (l1 != null && l2 != null) {
                if (l1.val < l2.val) {
                    cur.next = new ListNode(l1.val);
                    l1 = l1.next;
                } else {
                    cur.next = new ListNode(l2.val);
                    l2 = l2.next;
                }
                cur = cur.next;
            }
            cur.next = l1 == null ? l2 : l1;
            return head.next;
        }
    }

    //递归
    class Solution {
        public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
            if (l1 == null) {
                return l2;
            } else if (l2 == null) {
                return l1;
            } else if (l1.val < l2.val) {
                l1.next = mergeTwoLists(l1.next, l2);
                return l1;
            } else {
                l2.next = mergeTwoLists(l1, l2.next);
                return l2;
            }
        }
    }


}